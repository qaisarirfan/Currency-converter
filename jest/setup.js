jest.mock('react-native-reanimated', () => {
  const Reanimated = require('react-native-reanimated/mock');

  // The mock for `call` immediately calls the callback which is incorrect
  // So we override it with a no-op
  Reanimated.default.call = () => {};

  return Reanimated;
});

jest.mock('@react-navigation/native', () => {
  const actualNav = jest.requireActual('@react-navigation/native');
  return {
    ...actualNav,
    useFocusEffect: () => jest.fn(),
    useRoute: () => jest.fn(),
    useNavigation: () => ({
      navigation: () => jest.fn(),
      canGoBack: () => jest.fn(),
      goBack: () => jest.fn(),
      push: () => jest.fn(),
    }),
  };
});

jest.mock('react-native-bootsplash', () => ({
  hide: jest.fn(),
  show: jest.fn(),
}));

jest.mock('react-native-localize', () => ({
  getLocales: () => [
    { countryCode: 'GB', languageTag: 'en-GB', languageCode: 'en', isRTL: false },
    { countryCode: 'US', languageTag: 'en-US', languageCode: 'en', isRTL: false },
    { countryCode: 'FR', languageTag: 'fr-FR', languageCode: 'fr', isRTL: false },
  ],

  getNumberFormatSettings: () => ({
    decimalSeparator: '.',
    groupingSeparator: ',',
  }),

  getCalendar: () => 'gregorian', // or "japanese", "buddhist"
  getCountry: () => 'US', // the country code you want
  getCurrencies: () => ['USD', 'EUR'], // can be empty array
  getTemperatureUnit: () => 'celsius', // or "fahrenheit"
  getTimeZone: () => 'Europe/Paris', // the timezone you want
  uses24HourClock: () => true,
  usesMetricSystem: () => true,

  addEventListener: jest.fn(),
  removeEventListener: jest.fn(),
}));

jest.mock('react-native-keyboard-aware-scroll-view', () => {
  const KeyboardAwareScrollView = require('react-native').ScrollView;
  return { KeyboardAwareScrollView };
});

jest.mock('react-native-gesture-handler', () => {
  // eslint-disable-next-line global-require
  const View = require('react-native/Libraries/Components/View/View');
  return {
    Swipeable: View,
    DrawerLayout: View,
    State: {},
    ScrollView: View,
    Slider: View,
    Switch: View,
    TextInput: View,
    ToolbarAndroid: View,
    ViewPagerAndroid: View,
    DrawerLayoutAndroid: View,
    WebView: View,
    NativeViewGestureHandler: View,
    TapGestureHandler: View,
    FlingGestureHandler: View,
    ForceTouchGestureHandler: View,
    LongPressGestureHandler: View,
    PanGestureHandler: View,
    PinchGestureHandler: View,
    RotationGestureHandler: View,
    /* Buttons */
    RawButton: View,
    BaseButton: View,
    RectButton: View,
    BorderlessButton: View,
    /* Other */
    FlatList: View,
    gestureHandlerRootHOC: jest.fn(),
    Directions: {},
  };
});

jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper');
