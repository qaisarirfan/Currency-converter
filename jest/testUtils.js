import React from 'react';
import { I18nextProvider } from 'react-i18next';
import { Provider } from 'react-redux';
import { render } from 'react-native-testing-library';
import configureStore from 'redux-mock-store';

import { initialState as themesState } from '../src/redux/reducers/themes';
import configureI18next from '../src/services/i18n';

const INITIAL_STATE = {
  themes: themesState,
};
const mockStore = configureStore([]);
const store = mockStore(INITIAL_STATE);

const reduxRender = (ui, options) =>
  render(ui, {
    wrapper: ({ children }) => (
      <Provider store={store}>
        <I18nextProvider i18n={configureI18next()}>{children}</I18nextProvider>
      </Provider>
    ),
    ...options,
  });

// re-export everything for convenience
export * from 'react-native-testing-library';

// override render method
export { reduxRender as render, store };
