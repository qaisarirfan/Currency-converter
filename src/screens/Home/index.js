import React, { useEffect, useState } from 'react';
import { View, TouchableOpacity, Text, ActivityIndicator } from 'react-native';
import { KeyboardAwareScrollView } from 'react-native-keyboard-aware-scroll-view';
import { useNavigation } from '@react-navigation/native';
import { useDispatch, useSelector } from 'react-redux';
import { useTranslation } from 'react-i18next';
import cc from 'currency-codes';
import Entypo from 'react-native-vector-icons/Entypo';

import ConversionInput from '../../components/ConversionInput';
import HeaderBar from '../../components/HeaderBar';
import KeyboardSpacer from '../../components/KeyboardSpacer';
import Logo from '../../components/Logo';
import ReverseButton from '../../components/ReverseButton';

import {
  selectBaseCurrency,
  selectCurrencies,
  selectFavoriteCurrencies,
  selectQuoteCurrency,
  selectRatesLoader,
} from '../../redux/reducers/conversion/selectors';
import { selectStyleableTheme } from '../../redux/reducers/themes/selectors';
import { createLoadAction, swapCurrency } from '../../redux/reducers/conversion/actionCreators';

import themeStyles from './styles';

function Home() {
  const { push, navigate } = useNavigation();
  const { t } = useTranslation();
  const dispatch = useDispatch();

  const [, setScrollEnabled] = useState(false);
  const [value, setValue] = useState('100');

  const baseCurrency = useSelector(selectBaseCurrency);
  const favoriteCurrencies = useSelector(selectFavoriteCurrencies);
  const loader = useSelector(selectRatesLoader);
  const quoteCurrency = useSelector(selectQuoteCurrency);
  const rates = useSelector(selectCurrencies);
  const styleableTheme = useSelector(selectStyleableTheme);

  const styles = themeStyles(styleableTheme);

  const conversionRate = (rates && rates.find((rate) => rate.name === quoteCurrency)) || {};

  const favConversion =
    favoriteCurrencies ||
    [].map((currency) => (rates && rates.find((rate) => rate.name === currency)) || {});

  const rate = conversionRate?.rate || 0;
  const inputValue = parseFloat(value) || 0;
  let convertedValue = t('home.converting');
  if (!loader) {
    convertedValue = (rate * inputValue).toFixed(2);
  }

  useEffect(() => {
    dispatch(createLoadAction(baseCurrency));
  }, [baseCurrency]);

  return (
    <View style={styles.root} testID="welcome">
      <HeaderBar
        title={t('home.title')}
        isHeaderShow={false}
        rightContent={
          <TouchableOpacity testID="options_screen_button" onPress={() => push('Options')}>
            <Entypo name="cog" size={32} color="#fff" />
          </TouchableOpacity>
        }
      />
      <KeyboardAwareScrollView testID="home_screen" scrollEnabled behavior="padding">
        <View style={styles.content}>
          <Logo />
          <Text style={styles.textHeader}>{t('home.currency_converter')}</Text>
          <View style={styles.inputContainer}>
            <ConversionInput
              keyboardType="numeric"
              testID="conversion_input_1"
              value={value.toString()}
              text={baseCurrency}
              onChangeText={setValue}
              onButtonPress={() =>
                navigate('Options', {
                  screen: 'CurrencyList',
                  params: {
                    title: t('home.base_currency'),
                    isBaseCurrency: true,
                  },
                })
              }
            />
            <ConversionInput
              editable={false}
              testID="conversion_input_2"
              text={quoteCurrency}
              value={convertedValue}
              onButtonPress={() =>
                navigate('Options', {
                  screen: 'CurrencyList',
                  params: {
                    title: t('home.quote_currency'),
                    isBaseCurrency: false,
                  },
                })
              }
            />
            <ReverseButton
              text={t('home.reverse_currencies')}
              onPress={() => dispatch(swapCurrency())}
            />
            {loader ? <ActivityIndicator color="#fff" /> : null}
          </View>
          {!loader && (
            <View style={styles.listWrapper}>
              {favConversion.map((v) => (
                <View style={styles.list} key={`fav-${v.name}`}>
                  <Text>{cc.code(v.name).currency}</Text>
                  <Text>
                    {v.name} {(v.rate * inputValue).toFixed(2) || ''}
                  </Text>
                </View>
              ))}
            </View>
          )}
          <KeyboardSpacer onToggle={(visible) => setScrollEnabled(visible)} />
        </View>
      </KeyboardAwareScrollView>
    </View>
  );
}

export default Home;
