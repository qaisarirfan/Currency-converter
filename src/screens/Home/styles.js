import { Dimensions, StyleSheet } from 'react-native';

const screen = Dimensions.get('window');

// Home: Styling here
const styles = (theme) =>
  StyleSheet.create({
    root: {
      backgroundColor: theme[500],
      flex: 1,
    },
    header: {
      alignItems: 'flex-end',
      marginHorizontal: 20,
    },
    content: {
      paddingTop: screen.height * 0.04,
    },
    textHeader: {
      color: theme[50],
      fontSize: 30,
      fontWeight: 'bold',
      marginBottom: 20,
      textAlign: 'center',
    },
    listWrapper: {
      backgroundColor: 'white',
      padding: 12,
    },
    list: {
      borderBottomColor: theme[500],
      borderBottomWidth: StyleSheet.hairlineWidth,
      display: 'flex',
      flexDirection: 'row',
      justifyContent: 'space-between',
      paddingBottom: 6,
      paddingTop: 6,
    },
  });

export default styles;
