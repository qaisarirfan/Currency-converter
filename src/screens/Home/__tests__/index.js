import React from "react"
import {render, cleanup, store} from '../../../../jest/testUtils';
import Home from ".."

afterEach(() => {
  cleanup();
  store.clearActions();
});

describe("<Home />", () => {
  it("Does not explode", () => {
    const rendered = render(<Home />).toJSON();
    expect(rendered).toMatchSnapshot();
  })
})
