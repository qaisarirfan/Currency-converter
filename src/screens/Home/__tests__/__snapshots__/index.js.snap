// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> Does not explode 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#4f6d7a",
      "flex": 1,
    }
  }
  testID="welcome"
>
  <View
    style={
      Array [
        Object {},
        Object {
          "paddingBottom": 0,
          "paddingTop": 0,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "height": 20,
          },
          Object {
            "backgroundColor": "#36505d",
          },
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "display": "flex",
            "flexDirection": "row",
            "flexGrow": 1,
            "justifyContent": "space-between",
            "paddingBottom": 12,
            "paddingTop": 12,
          },
          Object {
            "paddingBottom": 0,
            "paddingTop": 0,
          },
        ]
      }
    >
      <View
        style={
          Object {
            "alignItems": "flex-start",
            "justifyContent": "center",
            "paddingLeft": 6,
            "width": 250,
          }
        }
      />
      <View
        style={
          Object {
            "alignItems": "flex-end",
            "justifyContent": "center",
            "paddingRight": 6,
            "width": 250,
          }
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID="options_screen_button"
        >
          <Text
            allowFontScaling={false}
            selectable={false}
            style={
              Array [
                Object {
                  "color": "#fff",
                  "fontSize": 32,
                },
                undefined,
                Object {
                  "fontFamily": "Entypo",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            ï…œ
          </Text>
        </View>
      </View>
    </View>
  </View>
  <RCTScrollView
    behavior="padding"
    scrollEnabled={true}
    testID="home_screen"
  >
    <View>
      <View
        style={
          Object {
            "paddingTop": 53.36,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "justifyContent": "center",
                "marginBottom": 40,
              },
            ]
          }
        >
          <Image
            resizeMode="contain"
            source={
              Object {
                "process": [Function],
              }
            }
            style={
              Object {
                "height": 300,
                "width": 300,
              }
            }
          />
        </View>
        <Text
          style={
            Object {
              "color": "#eaedef",
              "fontSize": 30,
              "fontWeight": "bold",
              "marginBottom": 20,
              "textAlign": "center",
            }
          }
        >
          Currency converter
        </Text>
        <View>
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "#FFF",
                  "borderColor": "#a7b6bd",
                  "borderRadius": 5,
                  "borderWidth": 0.5,
                  "flexDirection": "row",
                  "marginHorizontal": 20,
                  "marginVertical": 10,
                },
              ]
            }
          >
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "backgroundColor": "#FFF",
                  "borderBottomLeftRadius": 5,
                  "borderRightColor": "#cad3d7",
                  "borderRightWidth": 0.5,
                  "borderTopLeftRadius": 5,
                  "opacity": 1,
                  "padding": 15,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#000",
                    "fontSize": 18,
                    "fontWeight": "bold",
                  }
                }
              />
            </View>
            <TextInput
              keyboardType="numeric"
              onChangeText={[Function]}
              style={
                Object {
                  "color": "#00476E",
                  "flex": 1,
                  "fontSize": 16,
                  "padding": 10,
                }
              }
              testID="conversion_input_1"
              value="100"
            />
          </View>
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "#FFF",
                  "borderColor": "#a7b6bd",
                  "borderRadius": 5,
                  "borderWidth": 0.5,
                  "flexDirection": "row",
                  "marginHorizontal": 20,
                  "marginVertical": 10,
                },
                Object {
                  "backgroundColor": "#f0f0f0",
                },
              ]
            }
          >
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "backgroundColor": "#FFF",
                  "borderBottomLeftRadius": 5,
                  "borderRightColor": "#cad3d7",
                  "borderRightWidth": 0.5,
                  "borderTopLeftRadius": 5,
                  "opacity": 1,
                  "padding": 15,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#000",
                    "fontSize": 18,
                    "fontWeight": "bold",
                  }
                }
              />
            </View>
            <TextInput
              editable={false}
              style={
                Object {
                  "color": "#00476E",
                  "flex": 1,
                  "fontSize": 16,
                  "padding": 10,
                }
              }
              testID="conversion_input_2"
              value="0.00"
            />
          </View>
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "center",
                "marginVertical": 20,
                "opacity": 1,
              }
            }
            testID="reverse_button"
          >
            <Image
              resizeMode="contain"
              source={
                Object {
                  "process": [Function],
                }
              }
              style={
                Object {
                  "height": 20,
                  "marginRight": 10,
                  "width": 20,
                }
              }
            />
            <Text
              style={
                Object {
                  "color": "#fff",
                  "fontSize": 16,
                }
              }
            >
              Reverse currencies
            </Text>
          </View>
        </View>
        <View
          style={
            Object {
              "backgroundColor": "white",
              "padding": 12,
            }
          }
        />
        <View
          style={
            Array [
              Object {
                "bottom": 0,
                "left": 0,
                "right": 0,
              },
              Object {
                "height": 0,
              },
              Object {},
            ]
          }
        />
      </View>
    </View>
  </RCTScrollView>
</View>
`;
